modules/_labs/select/react/lib/SelectMenu.tsx

```diff
@@ -133,20 +189,152 @@ const MenuList = styled('ul')<Pick<SelectProps, 'error' | 'theme'>>(
   })
 );
 
+const generatePopperOptions = (
+  props: Pick<SelectMenuProps, 'isAutoFlipped' | 'isAutoFocused' | 'isFlipped' | 'menuRef'>
+) => {
+  const {isAutoFlipped, isAutoFocused, isFlipped, menuRef} = props;
+
+  let fallbackPlacements: Placement[] = [];
+  if (isAutoFlipped) {
+    fallbackPlacements = isFlipped ? ['bottom-start'] : ['top-start'];
+  }
+
+  const modifiers = [
+    {
+      name: 'flip',
+      options: {
+        fallbackPlacements,
+      },
+    },
+    {
+      name: 'offset',
+      options: {
+        offset: ({
+          placement,
+          reference,
+          popper,
+        }: {
+          placement: Placement;
+          reference: Rect;
+          popper: Rect;
+        }) => {
+          // Skid menu along the edge of the button to account
+          // for fractional x-positioning of the button (e.g.,
+          // if the button is in a horizontally-centered modal)
+          // and to ensure proper alignment between the button
+          // and the menu
+          const skidding = reference.x % 1 >= 0.5 ? 1 : 0;
+
+          // Displace menu towards the button to obscure the bottom
+          // edge of the button and to create a smooth visual
+          // connection between the button and the menu
+          const distance = -parseInt(borderRadius.m, 10);
+
+          return [skidding, distance];
+        },
+      },
+    },
+  ];
+
+  return {
+    modifiers,
+    onFirstUpdate: () => {
+      if (isAutoFocused && menuRef && menuRef.current) {
+        menuRef.current.focus();
+      }
+    },
+  };
+};
+
 const SelectMenu = (props: SelectMenuProps) => {
-  const {children, error, isAnimated, isHiding, menuRef, ...elemProps} = props;
-
-  return (
-    <Menu error={error} isAnimated={isAnimated} isHiding={isHiding}>
-      <MenuList error={error} ref={menuRef} role="listbox" tabIndex={-1} {...elemProps}>
-        {children}
-      </MenuList>
-    </Menu>
-  );
+  const {
+    buttonRef,
+    children,
+    error,
+    isAnimated,
+    isAutoFlipped,
+    isAutoFocused,
+    isFlipped,
+    isHidden,
+    isHiding,
+    menuRef,
+    onCloseOnEscape,
+    ...elemProps
+  } = props;
+
+  const popupRef = React.useRef<HTMLDivElement>(null);
+
+  const [width, setWidth] = useState(0);
+
+  const handleWidthChange = () => {
+    if (buttonRef.current) {
+      const newMenuWidth = buttonRef.current.clientWidth + 2 * buttonBorderWidth;
+      setWidth(newMenuWidth);
+    }
+  };
+
+  useEffect(() => {
```

Comment:
Should this be `useLayoutEffect` instead? `useEffect` will always have a 1 frame delay which will cause a "flickering" effect.

```suggestion
  useLayoutEffect(() => {
```

Reply:
I'm not seeing a flickering effect here. Are you seeing something on your end?

Reply:
Yup. _Technically_, it can flicker because the width state of the menu initializes at 0 and then gets calculated and set asynchronously.

Yet, I'm not sure if that flicker ever shows up through regular usage of a `Select`. We animate the menu in via a fade so if the menu width is 0 for one frame... it won't be perceptible.

Clicking through tests was the only way I could get it to show up. If you click on the visual testing story "Select States Menu On" you can see the 1 frame difference between `useEffect` and `useLayoutEffect`.

Here's me scrubbing through a timeline in Chrome (very slowly). You'll see that the menu is supposed to be open when the story loads but it takes one frame before it shows.

![Screen-Recording-2020-06-11-at-11 10 44-PM](https://user-images.githubusercontent.com/14299381/84470838-0843dd00-ac39-11ea-9402-d192c616c382.gif)

Here it is with `useLayoutEffect`, the menu paints synchronously to having its width set.

![Screen-Recording-2020-06-11-at-11 14 53-PM](https://user-images.githubusercontent.com/14299381/84471211-b9e30e00-ac39-11ea-9861-e6998187c9e0.gif)

Reply:
Ah, I didn't realize that was the flicker you were referring to. I had noticed the menu delay earlier when pulling up the Menu On story, but I didn't realize it was due to `useEffect`. @lychyi Thanks for the detailed explanation! I've replaced `useEffect` with `useLayoutEffect`.

Reply:
I updated the Menu to use `bottom` instead of `bottom-start` which means Popper could open before the width is set causing issues where the wrong position was chosen. I could see this in IE11 where the menu would open with a `0` width causing the Popper's left side to be in the middle of the button.

I fixed this issue by keeping the Popup closed until width was known. This would cause the issue @lychyi showed. This wouldn't be a problem normally as chances are a user could not open a menu faster than `useEffect` calculating width.

https://github.com/Workday/canvas-kit/pull/641/commits/0fa2671a701ba4e4a284afe4c07f72988fff7ab8 introduced an optimization that delays the calculation of width to when we want to open the menu, making it possible for IE to choose the wrong location. But the Popper doesn't actually "open" until width is non-zero, so this shouldn't happen in practice, but `useLayoutEffect` does prevent the extra frame.